sdk/src/builder.rs

```diff
@@ -864,9 +877,105 @@ impl Builder {
             }?;
         }
 
+        if !found_actions {
+            let mut actions = Actions::new();
+            self.add_auto_actions_assertions(&mut actions)?;
+
+            if !actions.actions().is_empty() {
+                claim.add_assertion(&actions)?;
+            }
+        }
+
         Ok(claim)
     }
 
+    /// Adds c2pa.created, c2pa.opened, and c2pa.placed actions for the specified [Actions][crate::assertions::Actions]
+    /// assertion if the condiitons are applicable as defined in the spec.
+    ///
+    /// This function takes into account the [Settings][crate::Settings]:
+    /// * `builder.auto_created_action`
+    /// * `builder.auto_opened_action`
+    /// * `builder.auto_placed_action`
+    fn add_auto_actions_assertions(&self, actions: &mut Actions) -> Result<()> {
+        // https://spec.c2pa.org/specifications/specifications/2.2/specs/C2PA_Specification.html#_mandatory_presence_of_at_least_one_actions_assertion
+        let auto_created =
+            settings::get_settings_value::<bool>("builder.auto_created_action.enabled")?;
+        let auto_opened =
+            settings::get_settings_value::<bool>("builder.auto_opened_action.enabled")?;
+        if auto_created || auto_opened {
+            let has_parent = self
```

Comment:
We need to base this not on the manifest defintion but on the generated ingredient hashed uri map. We need that in order to add the ingredient references to the actions. It probably makes sense to update all the actions here, not just the ones we are adding. But that can come later.

Reply:
With all of these we need to check if the action is already defined. The user may have added it and we don't want to add it twice.